### PR TITLE
mocktikv,tidb-server: implement reverse scan and make mock tikv default engine

### PR DESCRIPTION
**Pull request by tiancaiamao**
I ignore `ReverseScan` when writing the MVCCLevelDB last time, 
so we have trouble change to the new mock tikv store.
It should work now.
@disksing @shenli @zimulala

### CODE REVIEW

**Comment by zz-jason** [comment]
@zimulala PTAL

**Comment by tiancaiamao** [comment]
@disksing @coocood PTAL

File: store/tikv/mock-tikv/mvcc_leveldb.go [hunk 1]
```diff
@@ -395,10 +395,98 @@ func (mvcc *MVCCLevelDB) Scan(startKey, endKey []byte, limit int, startTS uint64
...
+	startTS  uint64
+	isoLevel kvrpcpb.IsolationLevel
+	currKey  []byte
+	entry    mvccEntry
+	pairs    []Pair
+}
+
+func (helper *reverseScanHelper) addLock(key []byte, lock *mvccLock) {
+	if bytes.Compare(key, helper.currKey) != 0 {
+		helper.finishEntry(key)
```
Comment (coocood) [inline store/tikv/mock-tikv/mvcc_leveldb.go]: This function does more than its name suggests.
I think move this logic to line 428 is better.

File: store/tikv/mock-tikv/mvcc_leveldb.go [hunk 2]
```diff
@@ -451,17 +454,11 @@ type reverseScanHelper struct {
 	pairs    []Pair
 }
 
-func (helper *reverseScanHelper) addLock(key []byte, lock *mvccLock) {
-	if bytes.Compare(key, helper.currKey) != 0 {
-		helper.finishEntry(key)
-	}
+func (helper *reverseScanHelper) addLock(lock *mvccLock) {
```
Comment (coocood) [inline store/tikv/mock-tikv/mvcc_leveldb.go]: Those two methods can be removed as its just one line of code.

File: store/tikv/mock-tikv/mvcc_leveldb.go [hunk 3]
```diff
@@ -395,10 +395,95 @@ func (mvcc *MVCCLevelDB) Scan(startKey, endKey []byte, limit int, startTS uint64
...
+	val, err := helper.entry.Get(helper.startTS, helper.isoLevel)
+	if len(val) != 0 || err != nil {
+		helper.pairs = append(helper.pairs, Pair{
+			Key:   helper.currKey,
+			Value: val,
+			Err:   err,
+		})
+	}
+	helper.entry = mvccEntry{}
+	helper.currKey = key
```
Comment (coocood) [inline store/tikv/mock-tikv/mvcc_leveldb.go]: I think this line better be moved to line 431 for the same reason.
It does more than the name suggests.
And the argument can be removed.

**Comment by tiancaiamao** [comment]
@disksing @coocood

**Comment by coocood** [comment]
LGTM

File: store/tikv/mock-tikv/mvcc_leveldb.go [hunk 4]
```diff
@@ -395,10 +395,87 @@ func (mvcc *MVCCLevelDB) Scan(startKey, endKey []byte, limit int, startTS uint64
...
+	for succ && len(helper.pairs) < limit {
+		key, ver, err := mvccDecode(iter.Key())
+		if err != nil {
+			break
+		}
+		if bytes.Compare(key, startKey) < 0 {
+			break
+		}
+
+		if bytes.Compare(key, helper.currKey) != 0 {
```
Comment (disksing) [inline store/tikv/mock-tikv/mvcc_leveldb.go]: bytes.Equal

File: store/tikv/mock-tikv/mvcc_leveldb.go [hunk 5]
```diff
@@ -395,10 +395,87 @@ func (mvcc *MVCCLevelDB) Scan(startKey, endKey []byte, limit int, startTS uint64
...
+			helper.finishEntry()
+			helper.currKey = key
+		}
+		if ver == lockVer {
+			var lock mvccLock
+			err = lock.UnmarshalBinary(iter.Value())
```
Comment (disksing) [inline store/tikv/mock-tikv/mvcc_leveldb.go]: check err.

File: store/tikv/mock-tikv/mvcc_leveldb.go [hunk 6]
```diff
@@ -395,10 +395,87 @@ func (mvcc *MVCCLevelDB) Scan(startKey, endKey []byte, limit int, startTS uint64
...
+			helper.entry.lock = &lock
+		} else {
+			var value mvccValue
+			err = value.UnmarshalBinary(iter.Value())
```
Comment (disksing) [inline store/tikv/mock-tikv/mvcc_leveldb.go]: check err.

File: store/tikv/mock-tikv/mvcc_leveldb.go [hunk 7]
```diff
@@ -395,10 +395,87 @@ func (mvcc *MVCCLevelDB) Scan(startKey, endKey []byte, limit int, startTS uint64
...
+	currKey  []byte
+	entry    mvccEntry
+	pairs    []Pair
+}
+
+func (helper *reverseScanHelper) finishEntry() {
+	reverse(helper.entry.values)
+	helper.entry.key = NewMvccKey(helper.currKey)
+	val, err := helper.entry.Get(helper.startTS, helper.isoLevel)
+	if len(val) != 0 || err != nil {
```
Comment (disksing) [inline store/tikv/mock-tikv/mvcc_leveldb.go]: Is it possible that both `val` and `err` are nil?

Comment (coocood) [reply store/tikv/mock-tikv/mvcc_leveldb.go]: It can be a deleted entry.

Comment (disksing) [reply store/tikv/mock-tikv/mvcc_leveldb.go]: Aha, you're right.

Comment (tiancaiamao) [reply store/tikv/mock-tikv/mvcc_leveldb.go]: I just follow the old logic in mvcc.go, if both `val` and `err` are nil, this entry doesn't exist and it doesn't need to append them to results.

File: store/tikv/mock-tikv/mvcc_leveldb.go (hunk 6 above)
Comment (tiancaiamao) [reply store/tikv/mock-tikv/mvcc_leveldb.go]: Done. In the next lines.

**Comment by disksing** [comment]
LGTM.

**Pull request merged**